docs/layer_util.js

```diff
@@ -226,43 +251,102 @@ function showOverlayLayer(overlay, index, map) {
+ *     finishes rendering
+ * @return {Function} A callback to be passed to
+ *     CompositeImageMapType.setTileCallback or
+ *     ee.MapLayerOverlay.addTileCallback
  */
-function resolveOnTilesFinished(layerDisplayData, resolve) {
-  if (layerDisplayData.tileCallbackId) {
-    layerDisplayData.overlay.removeTileCallback(
-        layerDisplayData.tileCallbackId);
-  }
-  layerDisplayData.tileCallbackId =
-      layerDisplayData.overlay.addTileCallback((tileEvent) => {
-        if (tileEvent.count === 0) {
-          if (resolve) {
-            // This is the first time we've finished loading, so inform caller.
-            resolve();
-            // Free up reference to resolve and make future redraws add a
-            // loading element.
-            resolve = null;
-          } else {
-            // Loading has finished for a pan/zoom-triggered load.
-            loadingElementFinished(mapContainerId);
-          }
-          layerDisplayData.loading = false;
-        } else if (!resolve && !layerDisplayData.loading) {
-          // We've started loading again after the first time completed (because
-          // of a pan/zoom of the map). Enable loading indicator.
-          layerDisplayData.loading = true;
-          addLoadingElement(mapContainerId);
-        }
-      });
+function createTileCallback(layerDisplayData, resolve) {
+  return (tileEvent) => {
+    if (tileEvent.count === 0) {
+      if (resolve) {
+        // This is the first time we've finished loading, so inform caller.
+        resolve();
+        // Free up reference to resolve and make future redraws add a
+        // loading element.
+        resolve = null;
+      } else {
+        // Loading has finished for a pan/zoom-triggered load.
+        loadingElementFinished(mapContainerId);
+      }
+      layerDisplayData.loading = false;
+    } else if (!resolve && !layerDisplayData.loading) {
+      // We've started loading again after the first time completed (because
+      // of a pan/zoom of the map). Enable loading indicator.
+      layerDisplayData.loading = true;
+      addLoadingElement(mapContainerId);
+    }
+  };
+}
+
+/**
+ * Displays a collection of tiles (given by the 'tile-urls' attribute of layer)
+ * on the map using a CompositeImageMapType to combine all tiles for a given
+ * map location into one tile.
+ * @param {google.maps.Map} map
+ * @param {Object} layer Data for layer coming from Firestore
+ * @return {Promise} Promise that resolves when images are all downloaded
+ */
+function addTileLayer(map, layer) {
+  const layerDisplayData = new LayerDisplayData(null, true);
+  layerArray[layer['index']] = layerDisplayData;
+  layerDisplayData.overlay = new CompositeImageMapType({
+    tileUrls: layer['tile-urls'],
+    tileSize: layer['tile-size'],
+    maxZoom: layer['maxZoom'],
+    opacity: layer['opacity'],
+  });
+  return createCompositeTilePromise(layerDisplayData, layer['index'], map);
```

Comment:
Can you remind me where we landed on if these layers were actually going to be multiple layers/multiple indices?

Reply:
They're one index and layer, that's one big reason we're using a custom CompositeImageMapType and not the standard ImageMapType :)

(retrying 429 errors is the other)

You can see how that's the case because toggleLayerOff didn't have to change, and similarly the indices of layers coming from Firestore aren't modified at all.